traits/trait_list_object.py

```diff
@@ -53,460 +88,750 @@ def __init__(self, index=0, removed=None, added=None):
+        added : value or list of values
+            Value or list of values that were added
+
+        """
+        is_trait_none = self.trait is None
+        is_name_items_none = self.name_items is None
+        if not hasattr(self, "trait") or is_trait_none or is_name_items_none:
             return
 
-        removed = self.copy()
-        list.clear(self)
+        object = self.object()
+        if object is None:
+            return
 
-        if self.name_items is not None:
-            self._send_trait_items_event(
-                self.name_items, TraitListEvent(0, removed)
-            )
+        # bug-for-bug conversion of parameters to TraitListEvent
+        if isinstance(index, slice):
+            if index.step in {1, None}:
+                index = min(index.start, index.stop)
+            else:
+                if added and not isinstance(added, list):
+                    added = [added]
+                if not isinstance(removed, list):
+                    removed = [removed]
+        else:
+            if removed is Undefined:
+                removed = []
+            elif not isinstance(removed, list):
+                removed = [removed]
+            if added is Undefined:
+                added = []
+            elif not isinstance(added, list):
+                added = [added]
+        event = TraitListEvent(index, removed, added)
+        items_event = self.trait.items_event()
+        object.trait_items_event(self.name_items, event, items_event)
 
-    def len_error(self, len):
-        """ Utility method that raises an error if length is incorrect. """
-        raise TraitError(
-            "The '%s' trait of %s instance must be %s, "
-            "but you attempted to change its length to %d element%s."
-            % (
-                self.name,
-                class_of(self.object()),
-                self.trait.full_info(self.object(), self.name, Undefined),
-                len,
-                "s"[len == 1:],
-            )
+    def __deepcopy__(self, memo):
+        """ Perform a deepcopy operation.
+
+        Notifiers are transient and should not be copied.
+        """
+        id_self = id(self)
+        if id_self in memo:
+            return memo[id_self]
+
+        memo[id_self] = result = TraitListObject(
+            self.trait,
+            lambda: None,
+            self.name,
+            [copy.deepcopy(x, memo) for x in self],
         )
 
+        return result
+
     def __getstate__(self):
+        """ Get the state of the object for serialization.
+
+        Notifiers are transient and should not be serialized.
+        """
         result = self.__dict__.copy()
         result.pop("object", None)
         result.pop("trait", None)
 
         return result
 
     def __setstate__(self, state):
+        """ Restore the state of the object after serialization.
+
+        Notifiers are transient and are restored to the empty list.
+        """
```

Comment:
Same here, should we call `super.__setstate__` to handle the notifiers, and to add `self.notifier ` back?

Reply:
Yes

Reply:
Unresolving this comment, the `self.notifier` is not restored:
```

    def test_list_of_lists_pickle_with_notifier(self):
        tl = TraitListObject(
            trait=List(),
            object=Foo(),
            name="foo",
            value=(),
        )

        self.assertEqual(
            tl.notifiers,
            [tl.notifier],
        )

        serialized = pickle.dumps(tl)

        tl_deserialized = pickle.loads(serialized)

        self.assertEqual(
            tl_deserialized.notifiers,
            [tl_deserialized.notifier]
        )

```

failure:
```
======================================================================
FAIL: test_list_of_lists_pickle_with_notifier (traits.tests.test_trait_list_object.TestTraitList)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/kchoi/Work/ETS/traits/traits/tests/test_trait_list_object.py", line 344, in test_list_of_lists_pickle_with_notifier
    [tl_deserialized.notifier]
AssertionError: Lists differ: [] != [<bound method TraitListObject.notifier of []>]

Second list contains 1 additional elements.
First extra element 0:
<bound method TraitListObject.notifier of []>

- []
+ [<bound method TraitListObject.notifier of []>]
```
I am not sure calling `super().__setstate__(state)` here is right: It sets all the states to `__dict__`. It might be easier to handle all state restoration without relying on parent class.

Reply:
(apologies for backtracking on the `super().__setstate__` call there).
Essentially, this change would do:
Replace:
```
super().__setstate__(state)
```
with
```
state["notifiers"] = [self.notifier]
```

Reply:
Made the change, adding this test to the suite.